Allow var / let as parameter names, but provide a warning, and fixit to add backticks.

sl opened the pull request:
## Notable Changes
- Allows var / let to be used as parameter names as there's no reason to disallow this now
- Added a diagnostic recommending that people escape var / let used as keywords in function declarations. No escaping is recommended at the call site as there is no possible ambiguity regarding the possible behavior of let / var there.
- Removed let and var from the warning regarding multiple specifiers
- Updated tests to reflect the new behavior, and added new tests ensuring the new warning is properly generated

Resolves [SR-10293](https://bugs.swift.org/browse/SR-10293).
This is resolved as there's no longer a need to change that fixit.

sl committed:
AST: Removed extraneous space from let/var as a parameter fixit.

theblixguy commented:
Looks like https://github.com/apple/swift/pull/24025

CodaFi commented:
If you wish to take over this bug, please see my notes on the other. In brief: please don’t change the diagnostic engine, but instead remove this diagnostic and make it do something useful by sticking in backquotes.

CodaFi commented on test/FixCode/fixits-var-let-parameter.swift:
```diff
@@ -0,0 +1,7 @@
+func foo(var arg: Int) {}
+func bar(let arg: Int) {}
+
+func baz(var arg: Int, var arg2: Int) {}
```
Going through a full transformation is a bit heavy.  A simple verifier test that checks you nuked whitespace should suffice

xedin replied:
We have a fix-it syntax in double braces which could be used to make sure that it removes enough whitespace.

xedin commented on lib/AST/DiagnosticEngine.cpp:
```diff
@@ -173,10 +173,12 @@ InFlightDiagnostic &InFlightDiagnostic::fixItRemove(SourceRange R) {
+
   if (extractCharAfter(SM, charRange.getEnd()) == ' ' &&
-      isspace(extractCharBefore(SM, charRange.getStart()))) {
+      (isspace(charBefore) || charBefore == '(')) {
```
I think this should be done at the diagnostic site - https://github.com/apple/swift/blob/master/lib/Parse/ParsePattern.cpp#L239

@CodaFi I think it might make sense to give the diagnostic here, maybe a bit re-worded, and suggest to either remove the keyword via note or add backticks as you suggested in JIRA since it's not really obvious that user really meant to use `var` or `let` as a label. WDYT?

CodaFi replied:
Actually, come to think of it, why don't we just allow this without a diagnostic?  `var` and `let` no longer mean anything in this position, and we already allow other keywords in this position:

```
func foo<T>(class x: T) -> T { 
  return x  
} 
```

sl replied:
I suspect that allowing this could cause a lot of confusion. Keywords like `class` really couldn't have another reasonable meaning in this context, so I'm guessing that's why they're allowed without backticks. Allowing the use of `var` here without backticks would probably make people think the parameter is a var binding locally when it's still a let binding. That said, @CodaFi  that is a good point, if we don't care about the possibility of confusion there, it might make sense to just allow those names. Thoughts @xedin?

xedin replied:
I'm guessing that let/var where explicitly prohibited because of the meaning they have, maybe if we were to all them it would make sense to have a warning which says that keyword `let`/`var` in that position is interpreted as a label?

CodaFi replied:
That seems reasonable.  It's pretty much always a mistake when this shows up.  You can suppress that diagnostic with the backpacks.

xedin replied:
Yeah, add a warning and note with a fix to add backticks is what I'm thinking.

sl replied:
Sounds good 👍

sl replied:
Update on this, because it makes sense now to give a warning when people add unnecessary backpacks at the call site for let / var, this can actually stay how it was.

xedin commented on test/FixCode/fixits-var-let-parameter.swift.result:
```diff
@@ -0,0 +1,7 @@
+func foo(arg: Int) {}
+func bar(arg: Int) {}
+
+func baz(arg: Int, arg2: Int) {}
```
Nit: Please add newline at the end of both files.

sl committed:
Parse / AST: Allow let / var as argument labels with a warning.

The diagnostic is now a warning and the new message alerts the user that
though it is valid to have let and var as argument label names,
they are interpreted as argument labels, not keywords.

sl commented:
Just pushed a change that allows let / var as argument label names but generates a warning when they're used. I'm going to have to update a lot of tests that rely on this being an error so don't run CI just yet, but I figured I'd put this up so that people can take a quick look at the implementation.

xedin commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -850,8 +850,8 @@ ERROR(parameter_specifier_as_attr_disallowed,none,
-ERROR(parameter_let_var_as_attr,none,
-      "'%0' as a parameter attribute is not allowed",
+WARNING(parameter_let_var_as_attr,none,
+      "'%0' in function paramters is interpreted as an argument label",
```
Typo `paramters` - `parameters`, maybe we should simplify this to - `'%0' in this position is interpreted as an argument label`?

xedin commented on lib/Parse/ParseExpr.cpp:
```diff
@@ -2109,7 +2109,7 @@ void Parser::parseOptionalArgumentLabel(Identifier &name, SourceLoc &loc) {
     auto escaped = Tok.isEscapedIdentifier();
     auto underscore = Tok.is(tok::kw__) || (escaped && text == "_");
-    if (escaped && !underscore && canBeArgumentLabel(text)) {
+    if (escaped && !underscore && shouldBeArgumentLabel(text)) {
```
I think it's okay to use `!Tok.isAny(tok::kw_let, tok::kw_var)` here instead of adding `shouldBeArgumentLabel` which sounds kind of out of place to me...

sl replied:
That was my initial instinct as well but unfortunately that won't work as the token being processed here isn't `tok::kw_let` or `tok::kw_var`. At this point, it's `tok::identifier` with the text "let" or "var". That's why the initial implementation called `StringExtras::canBeArgumentLabel` rather than `Token::canBeArgumentLabel`.

If you think it's cleaner to do the string comparison inline rather than adding shouldBeArgumentLabel, I can do that, but I found that a bit hard to read.

xedin replied:
Ah interesting, let's convert that method into an inline closure then with `requiresEscaping` name or something similar.

sl replied:
Just to clarify, do you mean something like this?
```
auto requiresEscaping = [](StringRef identifier) {
  return llvm::StringSwitch<bool>(identifier)
    .Case("let", false)
    .Case("var", false)
    .Default(canBeArgumentLabel(identifier));
};

if (escaped && !underscore && requiresEscaping(text)) {
  ...
}
```

xedin replied:
Yeah, that's exactly what I mean.

sl replied:
👍

xedin commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -213,7 +213,7 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
     
     // ('inout' | 'let' | 'var' | '__shared' | '__owned')?
```
`let` and `var` should be removed from this list.

xedin commented on lib/Parse/ParsePattern.cpp:
```diff
@@ -251,6 +247,13 @@ Parser::parseParameterClause(SourceLoc &leftParenLoc,
+      // generate a warning.
+      if (Tok.isAny(tok::kw_let, tok::kw_var)) {
+        diagnose(Tok, diag::parameter_let_var_as_attr, Tok.getText())
+          .fixItReplace(Tok.getLoc(), "`" + Tok.getText().str() + "`");
```
Just a general comment - you can run `@swift-ci build toolchain`, download the result and try in Xcode to make sure (I'm not saying that there shouldn't add a new test too) that everything works fine.

sl committed:
Test: Add and update tests for allowing var and let as argument labels

sl commented:
@swift-ci build toolchain

xedin commented:
@swift-ci build toolchain

sl commented:
@swift-ci please smoke test

xedin commented on test/decl/func/functions.swift:
```diff
@@ -123,13 +123,12 @@ func testObjCMethodCurry(_ a : ClassWithObjCMethod) -> (Int) -> () {
 
 // We used to crash on this.
-func rdar16786220(inout let c: Int) -> () { // expected-error {{parameter must not have multiple '__owned', 'inout', '__shared', 'var', or 'let' specifiers}} {{25-29=}}
+func rdar16786220(inout let c: Int) -> () { // expected-warning {{'let' in this position is interpreted as an argument label}} {{25-28=`let`}}
```
Let's add a test which would trigger original error and a couple where let is in the middle or last argument/parameter to cover more ground.

sl replied:
I added a few where the let is the middle or last parameter in unqualified_name.swift. If I add the following test here as well, would that be sufficient?

```swift
func multipleSpecifiers(inout __owned a: Int) {} // ...
```

xedin replied:
Yeah, that would be fine, thank you!

sl commented:
@swift-ci please smoke test

sl commented:
@swift-ci build toolchain

xedin commented on test/IDE/complete_escaped_keyword.swift:
```diff
@@ -27,7 +27,7 @@ enum MyEnum {
 // STATIC_PRIMARY-DAG: Decl[LocalVar]/Local:               self[#MyEnum.Type#]; name=self
 // STATIC_PRIMARY-DAG: Decl[EnumElement]/CurrNominal:      `class`({#struct: String#})[#MyEnum#]; name=`class`(struct: String)
-// STATIC_PRIMARY-DAG: Decl[EnumElement]/CurrNominal:      `let`({#`var`: String#})[#MyEnum#]; name=`let`(`var`: String)
+// STATIC_PRIMARY-DAG: Decl[EnumElement]/CurrNominal:      `let`({#var: String#})[#MyEnum#]; name=`let`(var: String)
```
Do you know where/how these results are generated? It would be great to keep backticks in place, otherwise that would generate a warning when picked.

sl replied:
I thought that this was generated completions at the call site and we don't produce warnings when var is used unescaped at the call site. So for example:

```swift
func foo(var: Int) {} // only this generates a warning
func bar(`var`: Int) {}

foo(var: 1)
bar(var: 1)
```

I might be wrong about these being call site completions though.

xedin replied:
Even if that's the case I think we should always produce a warning which suggests to escape the use of `var` or `let` regardless if it's in the declaration or a call.

xedin replied:
Actually let me step back a bit, I think we should have it consistent with other keywords in the call site, if e.g. `class` or `struct` don't suggest backticks `let` & `var` shouldn't too for consistency.

sl replied:
Final decision on this: We decided to not produce the warning at the call site.

It’s useful to give the warning in declarations because there’s the possible ambiguity regarding the behavior, but at the call site that ambiguity doesn’t exist. Providing that warning also makes it hard for people to write nice apis if they have a good reason to use var / let as a keyword. Additionally, that's consistent with the way we treat other similar keywords. For example the following produces no warnings:

```swift
func foo(`class`: Int) { print(`class`)}

foo(class: 3)
```

xedin replied:
+1

xedin reviewed (approved):
LGTM! Thank you, @sl!

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

sl commented:
@swift-ci please smoke test

sl commented:
@swift-ci please test source compatibility

sl commented:
@swift-ci please test source compatibility
